vllm/v1/engine/core.py

```diff
@@ -404,3 +426,85 @@ def process_output_socket(self, output_path: str):
                 socket.send_multipart((buffer, ), copy=False)
+
+
+class DPEngineCoreProc(EngineCoreProc):
+    """ZMQ-wrapper for running EngineCore in background process
+    in a data parallel context."""
+
+    def __init__(
+        self,
+        input_path: str,
+        output_path: str,
+        vllm_config: VllmConfig,
+        executor_class: type[Executor],
+        log_stats: bool,
+    ):
+        # Add process-specific prefix to stdout and stderr before
+        # we initialize the engine.
+        from multiprocessing import current_process
+        process_name = current_process().name
+        pid = os.getpid()
+        _add_prefix(sys.stdout, process_name, pid)
+        _add_prefix(sys.stderr, process_name, pid)
+
+        from vllm.platforms import current_platform
+        if current_platform.is_cuda_alike():
+            from vllm.platforms.cuda import device_id_to_physical_device_id
+            tp_size = vllm_config.parallel_config.tensor_parallel_size
+            dp_rank = vllm_config.parallel_config.data_parallel_rank
+            os.environ["CUDA_VISIBLE_DEVICES"] = ",".join(
+                str(device_id_to_physical_device_id(i))
+                for i in range(dp_rank * tp_size, (dp_rank + 1) * tp_size))
+
+        dp_size = vllm_config.parallel_config.data_parallel_size
+        self.dp_group = None if dp_size <= 1 else (
+            vllm_config.parallel_config.stateless_init_dp_group())
+
+        # Initialize the engine after setting up environment.
+        super().__init__(input_path, output_path, vllm_config, executor_class,
+                         log_stats)
+
+        self.counter = 0
+
+    def run_busy_loop(self):
+        """Core busy loop of the EngineCore for data parallel case."""
+
+        # Loop until process is sent a SIGINT or SIGTERM
+        while True:
+            # 1) Poll the input queue until there is work to do.
+            self._process_input_queue()
+
+            local_unfinished_reqs = self.scheduler.has_unfinished_requests()
+            if local_unfinished_reqs:
+                # 2) Step the engine core.
+                self._process_engine_step()
+
+                # Check if we have now finished all requests.
+                local_unfinished_reqs = (
+                    self.scheduler.has_unfinished_requests())
+            else:
+                assert self.global_unfinished_reqs
+                # There must be unfinished requests in DP peers, run a
+                # dummy forward pass.
+                self.execute_dummy_batch()
+
+            # 3) All-reduce operation to determine global unfinished reqs.
+            self.global_unfinished_reqs = self._has_global_unfinished_reqs(
+                local_unfinished_reqs)
+
+            if not self.global_unfinished_reqs:
+                # Notify client that we are pausing the loop.
+                self.output_queue.put_nowait(
+                    EngineCoreOutputs(engine_paused=True))
```

Comment:
When we pause the loop, does that just mean that we're spinning in `self._process_input_queue()`? And in effect this is just telling the client to send us a `start_dp_msg` if needed?

Reply:
Yes, though not actually spinning because it blocks on the input socket. Really it's saying that the engine_busy_loop is going to block wait on the input queue (which is fed from the input socket by `self._process_input_queue()` once any new messages arrive.